### ISSUE TITLE
The training is not working

### ISSUE THREAD

**Issue by huutuan1705** [Lightning-AI/pytorch-lightning#21463]
### Bug description

I have a problem. I tried to convert my old code from version 1.7.3 to 2.x, but the training process is not working. I see that because my code's result in 1.7.3 was successful, but with 2.0.0, the result at each epoch always was 0. What were I wrong?

My old code:
`class Model(pl.LightningModule):
    def __init__(self):

        super().__init__()
        self.opts = opts
        self.clip, _ = clip.load('ViT-B/32', device=self.device)
        self.clip.apply(freeze_all_but_bn)
        # Prompt Engineering
        self.sk_prompt = nn.Parameter(torch.randn(self.opts.n_prompts, self.opts.prompt_dim))
        self.img_prompt = nn.Parameter(torch.randn(self.opts.n_prompts, self.opts.prompt_dim))
        self.distance_fn = lambda x, y: 1.0 - F.cosine_similarity(x, y)
        self.loss_fn = nn.TripletMarginWithDistanceLoss(
            distance_function=self.distance_fn, margin=0.2)
        self.best_metric = -1e3

    def configure_optimizers(self):
        optimizer = torch.optim.Adam([
            {'params': self.clip.parameters(), 'lr': self.opts.clip_LN_lr},
            {'params': [self.sk_prompt] + [self.img_prompt], 'lr': self.opts.prompt_lr}])
        return optimizer

    def forward(self, data, dtype='image'):
        if dtype == 'image':
            feat = self.clip.encode_image(
                data, self.img_prompt.expand(data.shape[0], -1, -1))
        else:
            feat = self.clip.encode_image(
                data, self.sk_prompt.expand(data.shape[0], -1, -1))
        return feat

    def training_step(self, batch, batch_idx):
        sk_tensor, img_tensor, neg_tensor, category = batch[:4]
        img_feat = self.forward(img_tensor, dtype='image')
        sk_feat = self.forward(sk_tensor, dtype='sketch')
        neg_feat = self.forward(neg_tensor, dtype='image')

        loss = self.loss_fn(sk_feat, img_feat, neg_feat)
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        sk_tensor, img_tensor, neg_tensor, category = batch[:4]
        img_feat = self.forward(img_tensor, dtype='image')
        sk_feat = self.forward(sk_tensor, dtype='sketch')
        neg_feat = self.forward(neg_tensor, dtype='image')

        loss = self.loss_fn(sk_feat, img_feat, neg_feat)
        self.log('val_loss', loss)
        return sk_feat, img_feat, category

    def validation_epoch_end(self, val_step_outputs):
        Len = len(val_step_outputs)
        if Len == 0:
            return
        query_feat_all = torch.cat([val_step_outputs[i][0] for i in range(Len)])
        gallery_feat_all = torch.cat([val_step_outputs[i][1] for i in range(Len)])
        all_category = np.array(sum([list(val_step_outputs[i][2]) for i in range(Len)], []))

        ## mAP category-level SBIR Metrics
        gallery = gallery_feat_all
        ap = torch.zeros(len(query_feat_all))
        for idx, sk_feat in enumerate(query_feat_all):
            category = all_category[idx]
            distance = -1*self.distance_fn(sk_feat.unsqueeze(0), gallery)
            target = torch.zeros(len(gallery), dtype=torch.bool)
            target[np.where(all_category == category)] = True
            ap[idx] = retrieval_average_precision(distance.cpu(), target.cpu())
        
        mAP = torch.mean(ap)
        self.log('mAP', mAP)
        if self.global_step > 0:
            self.best_metric = self.best_metric if  (self.best_metric > mAP.item()) else mAP.item()
        print ('mAP: {}, Best mAP: {}'.format(mAP.item(), self.best_metric))
`
The result of old code:
`Sanity Checking DataLoader 0: 100%|███████████████| 2/2 [00:02<00:00,  1.29s/it]mAP: 1.0, Best mAP: -1000.0
Epoch 4: 100%|██████████| 591/591 [13:47<00:00,  1.40s/it, loss=0.0237, v_num=0]mAP: 0.682223379611969, Best mAP: 0.682223379611969
`

My new code is:

`class Model(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.opts = opts
        self.clip, _ = clip.load('ViT-B/32', device=self.device)
        self.clip.apply(freeze_all_but_bn)

        # Prompt Engineering
        self.sk_prompt = nn.Parameter(torch.randn(self.opts.n_prompts, self.opts.prompt_dim))
        self.img_prompt = nn.Parameter(torch.randn(self.opts.n_prompts, self.opts.prompt_dim))

        self.distance_fn = lambda x, y: 1.0 - F.cosine_similarity(x, y)
        self.loss_fn = nn.TripletMarginWithDistanceLoss(
            distance_function=self.distance_fn, margin=0.2)
        self.val_step_outputs = []
        self.train_output = []
        self.best_metric = -1e3

    def configure_optimizers(self):
        optimizer = torch.optim.Adam([
            {'params': self.clip.parameters(), 'lr': self.opts.clip_LN_lr},
            {'params': [self.sk_prompt] + [self.img_prompt], 'lr': self.opts.prompt_lr}])
        return optimizer

    def forward(self, data, dtype='image'):
        if dtype == 'image':
            feat = self.clip.encode_image(
                data, self.img_prompt.expand(data.shape[0], -1, -1))
        else:
            feat = self.clip.encode_image(
                data, self.sk_prompt.expand(data.shape[0], -1, -1))
        return feat

    def training_step(self, batch, batch_idx):
        sk_tensor, img_tensor, neg_tensor, category = batch[:4]
        img_feat = self.forward(img_tensor, dtype='image')
        sk_feat = self.forward(sk_tensor, dtype='sketch')
        neg_feat = self.forward(neg_tensor, dtype='image')

        loss = self.loss_fn(sk_feat, img_feat, neg_feat)
        self.log('train_loss', loss)
        self.train_output.append(loss.item())
        
        return loss
    
    def on_train_epoch_end(self):
        self.train_output.clear()

    def validation_step(self, batch, batch_idx):
        sk_tensor, img_tensor, neg_tensor, category = batch[:4]
        img_feat = self.forward(img_tensor, dtype='image')
        sk_feat = self.forward(sk_tensor, dtype='sketch')
        neg_feat = self.forward(neg_tensor, dtype='image')

        loss = self.loss_fn(sk_feat, img_feat, neg_feat)
        self.log('val_loss', loss)
        self.val_step_outputs.append({
            "sk_feat": sk_feat,
            "img_feat": img_feat,
            "category": category
        })
        

    def on_validation_epoch_end(self):
        val_step_outputs = self.val_step_outputs
        Len = len(val_step_outputs)
        if Len == 0:
            return
        query_feat_all = torch.cat([val_step_outputs[i]["sk_feat"] for i in range(Len)])
        gallery_feat_all = torch.cat([val_step_outputs[i]["img_feat"] for i in range(Len)])
        all_category = np.array(sum([list(val_step_outputs[i]["category"]) for i in range(Len)], []))

        ## mAP category-level SBIR Metrics
        gallery = gallery_feat_all
        ap = torch.zeros(len(query_feat_all))
        pr = torch.zeros(len(query_feat_all))
        top_k = 200
        
        for idx, sk_feat in enumerate(query_feat_all):
            category = all_category[idx]
            distance = -1*self.distance_fn(sk_feat.unsqueeze(0), gallery)
            
            top_k_actual = min(top_k, len(gallery)) 
            top_values, top_indices = torch.topk(distance, top_k_actual, largest=True)
            
            target = torch.zeros(len(gallery), dtype=torch.bool, device=device)
            target[np.where(all_category == category)] = True
            
            ap[idx] = retrieval_average_precision(distance.cpu(), target.cpu(), top_k = top_k_actual )
            
        mAP = torch.mean(ap)
        mpr = torch.mean(pr)
        self.log('mAP', mAP, batch_size=1)
        if self.global_step > 0:
            self.best_metric = self.best_metric if  (self.best_metric > mAP.item()) else mAP.item()
        print ('mAP@200: {}, p@200: {}, Best mAP: {}'.format(mAP.item(), mpr.item(), self.best_metric))
        self.val_step_outputs.clear()`

And its result is:
`
/usr/local/lib/python3.12/dist-packages/torchmetrics/utilities/prints.py:70: 
FutureWarning: Importing `retrieval_average_precision` from 
`torchmetrics.functional` was deprecated and will be removed in 2.0. Import 
`retrieval_average_precision` from `torchmetrics.retrieval` instead.
  _future_warning(

mAP@200: 0.0, p@200: 0.0, Best mAP: -1000.0
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/loops/fit_loop.py:534:
Found 227 module(s) in eval mode at the start of training. This may lead to 
unexpected behavior during training. If this is intentional, you can ignore this
warning.

Epoch 1/49 ━━━━━━━━━━━━━━━━━━━━━ 491/491 0:12:21 • 0:00:00 0.66it/s v_num: 0.000
mAP@200: 0.0, p@200: 0.0, Best mAP: 0.0
`

Can anyone help me, please? If can't fix this problem, how to implement mAP@200 and precision@200 in version 1.7.3 manually?
Thanks so much.

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

cc @ethanwharris

**Comment by huutuan1705** [comment]
I fixed it. I think that `retrieval_average_precision` function does not receive tensors, which are negative value. The line `distance = -1*self.distance_fn(sk_feat.unsqueeze(0), gallery)` make distance become a negative tensor, so I just change it into `distance = 1 - self.distance_fn(sk_feat.unsqueeze(0), gallery)`

**Comment by bhimrazy** [comment]
Since the problem is resolved, I’m closing this issue. Please feel free to reopen or create a new one if needed. Thanks!
